index.d.ts

```diff
@@ -1,255 +1,234 @@
-
-	interface UriOptions {
-		uri: string;
-	}
-	interface UrlOptions {
-		url: string;
-	}
-	export type RequiredUriUrl = UriOptions | UrlOptions;
-
-	interface OptionalUriUrl {
-		uri?: string;
-		url?: string;
-	}
+  export interface RequestAPI<TRequest extends Request,
+    TOptions extends CoreOptions,
+    TUriUrlOptions> {
+
+    defaults(options: TOptions): RequestAPI<TRequest, TOptions, RequiredUriUrl>;
+    defaults(options: RequiredUriUrl & TOptions): DefaultUriUrlRequestApi<TRequest, TOptions, OptionalUriUrl>;
+
+    (uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    (uri: string, callback?: RequestCallback): TRequest;
+    (options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    get(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    get(uri: string, callback?: RequestCallback): TRequest;
+    get(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    post(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    post(uri: string, callback?: RequestCallback): TRequest;
+    post(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    put(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    put(uri: string, callback?: RequestCallback): TRequest;
+    put(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    head(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    head(uri: string, callback?: RequestCallback): TRequest;
+    head(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    patch(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    patch(uri: string, callback?: RequestCallback): TRequest;
+    patch(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    del(uri: string, options?: TOptions, callback?: RequestCallback): TRequest;
+    del(uri: string, callback?: RequestCallback): TRequest;
+    del(options: TUriUrlOptions & TOptions, callback?: RequestCallback): TRequest;
+
+    forever(agentOptions: any, optionsArg: any): TRequest;
+    jar(): CookieJar;
+    cookie(str: string): Cookie;
+
+    initParams: any;
+    debug: boolean;
+  }
+
+  interface DefaultUriUrlRequestApi<TRequest extends Request,
+    TOptions extends CoreOptions,
+    TUriUrlOptions>	extends RequestAPI<TRequest, TOptions, TUriUrlOptions> {
+
+    defaults(options: TOptions): DefaultUriUrlRequestApi<TRequest, TOptions, OptionalUriUrl>;
+    defaults(options: RequiredUriUrl & TOptions): DefaultUriUrlRequestApi<TRequest, TOptions, OptionalUriUrl>;
+    (): TRequest;
+    get(): TRequest;
+    post(): TRequest;
+    put(): TRequest;
+    head(): TRequest;
+    patch(): TRequest;
+    del(): TRequest;
+  }
+
+  interface CoreOptions {
+    baseUrl?: string;
+    callback?: (error: any, response: IncomingMessage, body: any) => void;
+    jar?: any; // CookieJar
```

Comment:
You are right, I just took you're declarations and imported blakeembrey's `typed-tough-cookie`.
I did not searched for every instance of Cookiejar. I'll fix it in 2 min